README.md
# Longhand
A parser for longhand strings.

## Installation
```
$ go get github.com/christianraza/longhand
```

## Usage
```
package main

import (
    "fmt"
    "log"

    "github.com/christianraza/longhand"
)

func main() {
    i, err := longhand.ParseLonghand("Forty-two")
    if err != nil {
        log.Fatal(err)
    }
    fmt.Printf("Forty-two as an integer: %d\n", i)
    i64, err := longhand.ParseLonghand64("Forty-two")
    if err != nil {
        log.Fatal(err)
    }
    fmt.Printf("Forty-two as a 64-bit integer: %d\n", i64)
}
```

## Supported Styles
##### Standard
`Five thousand two hundred and thirty-three` = `5233`
##### Years
`Nineteen thirty` = `1930`
##### Repetition
`Thirty thirty` = `3030`
##### Notes on style
What all this means is that something like `nineteen thirty thirty hundred` is valid input and will be parsed as `19303000`. `nineteen thirty thirty hundred` is not a common way of writing `19303000` but it is a possibility. So this, as well as many other variations, are supported styles.

## "Valid" Longhand
##### Examples of valid longhand
`Three` as opposed to `3`  
`Thirty-three` as opposed to `33`  
`Three hundred and thirty-three` as opposed to `333`
##### Examples of invalid longhand
`Thre` as opposed to `Three`  
`Thirty_three` as opposed to `Thirty-three`  
`3 hundred and thirty-three` as opposed to `Three hundred and thirty-three`
##### Consequences of invalid longhand
There are dire dire consequences for invalid longhand!  
:fearful: ...  
Just kidding :smile: invalid longhand just returns zero and an error.
##### "Valid"?
You likely noticed the quotes around valid in the title of this section, that's because the way people interpret numerals sometimes depends on location and preference. I'm hard pressed to say that there is a valid way of arranging numerals, and some of this can be seen in the section on styles. I had to make compromises here and there, as is usual when writing tools or software, but generally these functions should do something reasonable so have fun! :smiley:

CHANGELOG.md
# v1.0.0
Longhand is live!


longhand_test.go
package longhand

import (
	"testing"
)

// ParseLonghand tests
func TestLonghandEmpty(t *testing.T) {
	i, err := ParseLonghand("")
	if err == nil {
		t.Fatal(err)
	}
	var expecting int = 0
	if i != expecting {
		t.Fatalf("Expected: %d Got: %d", expecting, i)
	}
}

func TestLonghandSimple(t *testing.T) {
	i, err := ParseLonghand("two")
	if err != nil {
		t.Fatal(err)
	}
	var expecting int = 2
	if i != expecting {
		t.Fatalf("Expected: %d Got: %d", expecting, i)
	}
}

func TestLonghandComplex(t *testing.T) {
	i, err := ParseLonghand("seven thousand two hundred thirty one")
	if err != nil {
		t.Fatal(err)
	}
	var expecting int = 7231
	if i != expecting {
		t.Fatalf("Expected: %d Got: %d", expecting, i)
	}
}

func TestLonghandRepetition(t *testing.T) {
	i, err := ParseLonghand("thirty thirty twenty-one one one")
	if err != nil {
		t.Fatal(err)
	}
	var expecting int = 30302111
	if i != expecting {
		t.Fatalf("Expected: %d Got: %d", expecting, i)
	}
}

func TestLonghandYear(t *testing.T) {
	i, err := ParseLonghand("nineteen twenty")
	if err != nil {
		t.Fatal(err)
	}
	var expecting int = 1920
	if i != expecting {
		t.Fatalf("Expected: %d Got: %d", expecting, i)
	}
}

func TestLonghandAlternativeSpellingDash(t *testing.T) {
	i, err := ParseLonghand("seven thousand two hundred thirty-one")
	if err != nil {
		t.Fatal(err)
	}
	var expecting int = 7231
	if i != expecting {
		t.Fatalf("Expected: %d Got: %d", expecting, i)
	}
}

func TestLonghandAlternativeSpellingAnd(t *testing.T) {
	i, err := ParseLonghand("seven thousand two hundred and thirty one")
	if err != nil {
		t.Fatal(err)
	}
	var expecting int = 7231
	if i != expecting {
		t.Fatalf("Expected: %d Got: %d", expecting, i)
	}
}

func TestLonghandAlternativeSpellingDashAnd(t *testing.T) {
	i, err := ParseLonghand("seven thousand two hundred and thirty one")
	if err != nil {
		t.Fatal(err)
	}
	var expecting int = 7231
	if i != expecting {
		t.Fatalf("Expected: %d Got: %d", expecting, i)
	}
}

func TestLonghandNegative(t *testing.T) {
	i, err := ParseLonghand("negative seven thousand two hundred thirty one")
	if err != nil {
		t.Fatal(err)
	}
	var expecting int = -7231
	if i != expecting {
		t.Fatalf("Expected: %d Got: %d", expecting, i)
	}
}

func TestLonghandNegativeEdgeCase(t *testing.T) {
	i, err := ParseLonghand(`negative two billion one hundred forty seven million
						     four hundred eighty three thousand six hundred forty eight`)
	if err != nil {
		t.Fatal(err)
	}
	var expecting int = -2147483648
	if i != expecting {
		t.Fatalf("Expected: %d Got: %d", expecting, i)
	}
}

func TestLonghandPositiveEdgeCase(t *testing.T) {
	i, err := ParseLonghand(`two billion one hundred forty seven million
						     four hundred eighty three thousand six hundred forty seven`)
	if err != nil {
		t.Fatal(err)
	}
	var expecting int = 2147483647
	if i != expecting {
		t.Fatalf("Expected: %d Got: %d", expecting, i)
	}
}

func TestLonghandTypo(t *testing.T) {
	i, err := ParseLonghand("Three hunded twenty four thousand")
	if err == nil {
		t.Fatal(err)
	}
	var expecting int = 0
	if i != expecting {
		t.Fatalf("Expected: %d Got: %d", expecting, i)
	}
}

func TestLonghandMix(t *testing.T) {
	i, err := ParseLonghand("five 1000 one hundred eighty seven")
	if err == nil {
		t.Fatal(err)
	}
	var expecting int = 0
	if i != expecting {
		t.Fatalf("Expected: %d Got: %d", expecting, i)
	}
}

func TestLonghandMixLarge(t *testing.T) {
	i, err := ParseLonghand(`two billion one hundred forty seven million
						     four hundred eighty three thousand 647`)
	if err == nil {
		t.Fatal(err)
	}
	var expecting int = 0
	if i != expecting {
		t.Fatalf("Expected: %d Got: %d", expecting, i)
	}
}

func TestLonghandNegativeOverflow(t *testing.T) {
	i, err := ParseLonghand(`negative two billion one hundred forty seven million
							 four hundred eighty three thousand six hundred forty nine`)
	if err == nil {
		t.Fatal(err)
	}
	var expecting int = 0
	if i != expecting {
		t.Fatalf("Expected: %d Got: %d", expecting, i)
	}
}

func TestLonghandPositiveOverflow(t *testing.T) {
	i, err := ParseLonghand(`two billion one hundred forty seven million
						     four hundred eighty three thousand six hundred forty eight`)
	if err == nil {
		t.Fatal(err)
	}
	var expecting int = 0
	if i != expecting {
		t.Fatalf("Expected: %d Got: %d", expecting, i)
	}
}

// ParseLonghand64 tests
func TestLonghand64Empty(t *testing.T) {
	i64, err := ParseLonghand64("")
	if err == nil {
		t.Fatal(err)
	}
	var expecting int64 = 0
	if i64 != expecting {
		t.Fatalf("Expected: %d Got: %d", expecting, i64)
	}
}

func TestLonghand64Complex(t *testing.T) {
	i64, err := ParseLonghand64("seven thousand two hundred thirty one")
	if err != nil {
		t.Fatal(err)
	}
	var expecting int64 = 7231
	if i64 != expecting {
		t.Fatalf("Expected: %d Got: %d", expecting, i64)
	}
}

func TestLonghand64Repetition(t *testing.T) {
	i64, err := ParseLonghand64("thirty thirty twenty-one one one")
	if err != nil {
		t.Fatal(err)
	}
	var expecting int64 = 30302111
	if i64 != expecting {
		t.Fatalf("Expected: %d Got: %d", expecting, i64)
	}
}

func TestLonghand64Year(t *testing.T) {
	i64, err := ParseLonghand64("nineteen twenty")
	if err != nil {
		t.Fatal(err)
	}
	var expecting int64 = 1920
	if i64 != expecting {
		t.Fatalf("Expected: %d Got: %d", expecting, i64)
	}
}

func TestLonghand64AlternativeSpellingDash(t *testing.T) {
	i64, err := ParseLonghand64("seven thousand two hundred thirty-one")
	if err != nil {
		t.Fatal(err)
	}
	var expecting int64 = 7231
	if i64 != expecting {
		t.Fatalf("Expected: %d Got: %d", expecting, i64)
	}
}

func TestLonghand64AlternativeSpellingAnd(t *testing.T) {
	i64, err := ParseLonghand64("seven thousand two hundred and thirty one")
	if err != nil {
		t.Fatal(err)
	}
	var expecting int64 = 7231
	if i64 != expecting {
		t.Fatalf("Expected: %d Got: %d", expecting, i64)
	}
}

func TestLonghand64AlternativeSpellingDashAnd(t *testing.T) {
	i64, err := ParseLonghand64("seven thousand two hundred and thirty-one")
	if err != nil {
		t.Fatal(err)
	}
	var expecting int64 = 7231
	if i64 != expecting {
		t.Fatalf("Expected: %d Got: %d", expecting, i64)
	}
}

func TestLonghand64Negative(t *testing.T) {
	i64, err := ParseLonghand64("negative seven thousand two hundred thirty one")
	if err != nil {
		t.Fatal(err)
	}
	var expecting int64 = -7231
	if i64 != expecting {
		t.Fatalf("Expected: %d Got: %d", expecting, i64)
	}
}

func TestLonghand64NegativeEdgeCase(t *testing.T) {
	i64, err := ParseLonghand64(`negative nine quintillion two hundred twenty three quadrillion
								 three hundred seventy two trillion thirty six billion
								 eight hundred fifty four million seven hundred seventy
								 five thousand eight hundred eight`)
	if err != nil {
		t.Fatal(err)
	}
	var expecting int64 = -9223372036854775808
	if i64 != expecting {
		t.Fatalf("Expected: %d Got: %d", expecting, i64)
	}
}

func TestLonghand64PositiveEdgeCase(t *testing.T) {
	i64, err := ParseLonghand64(`nine quintillion two hundred twenty three quadrillion
								 three hundred seventy two trillion thirty six billion
								 eight hundred fifty four million seven hundred seventy
								 five thousand eight hundred seven`)
	if err != nil {
		t.Fatal(err)
	}
	var expecting int64 = 9223372036854775807
	if i64 != expecting {
		t.Fatalf("Expected: %d Got: %d", expecting, i64)
	}
}

func TestLonghand64Typo(t *testing.T) {
	i64, err := ParseLonghand64("Three hunded twenty four thousand")
	if err == nil {
		t.Fatal(err)
	}
	var expecting int64 = 0
	if i64 != expecting {
		t.Fatalf("Expected: %d Got: %d", expecting, i64)
	}
}

func TestLonghand64Mix(t *testing.T) {
	i64, err := ParseLonghand64("five 1000 one hundred eighty seven")
	if err == nil {
		t.Fatal(err)
	}
	var expecting int64 = 0
	if i64 != expecting {
		t.Fatalf("Expected: %d Got: %d", expecting, i64)
	}
}

func TestLonghand64MixLarge(t *testing.T) {
	i64, err := ParseLonghand64(`nine quintillion two hundred twenty three quadrillion
								 three hundred seventy two trillion thirty six billion
								 eight hundred fifty four million seven hundred seventy
								 five 1807`)
	if err == nil {
		t.Fatal(err)
	}
	var expecting int64 = 0
	if i64 != expecting {
		t.Fatalf("Expected: %d Got: %d", expecting, i64)
	}
}

func TestLonghand64NegativeOverflow(t *testing.T) {
	i64, err := ParseLonghand64(`negative nine quintillion two hundred twenty three quadrillion
							   	 three hundred seventy two trillion thirty six billion
							   	 eight hundred fifty four million seven hundred seventy
							     five thousand eight hundred nine`)
	if err == nil {
		t.Fatal(err)
	}
	var expecting int64 = 0
	if i64 != expecting {
		t.Fatalf("Expected: %d Got: %d", expecting, i64)
	}
}

func TestLonghand64PositiveOverflow(t *testing.T) {
	i64, err := ParseLonghand64(`nine quintillion two hundred twenty three quadrillion
							   	 three hundred seventy two trillion thirty six billion
							     eight hundred fifty four million seven hundred seventy
							     five thousand eight hundred eight`)
	if err == nil {
		t.Fatal(err)
	}
	var expecting int64 = 0
	if i64 != expecting {
		t.Fatalf("Expected: %d Got: %d", expecting, i64)
	}
}

func TestLonghand64NegativeWraparoundOverflow(t *testing.T) {
	i64, err := ParseLonghand64(`negative twenty quintillion two hundred twenty three quadrillion
							   	 three hundred seventy two trillion thirty six billion
							   	 eight hundred fifty four million seven hundred seventy
							   	 five thousand eight hundred nine`)
	if err == nil {
		t.Fatal(err)
	}
	var expecting int64 = 0
	if i64 != expecting {
		t.Fatalf("Expected: %d Got: %d", expecting, i64)
	}
}

func TestLonghand64PositiveWraparoundOverflow(t *testing.T) {
	i64, err := ParseLonghand64(`twenty quintillion two hundred twenty three quadrillion
							   	 three hundred seventy two trillion thirty six billion
							   	 eight hundred fifty four million seven hundred seventy
							   	 five thousand eight hundred eight`)
	if err == nil {
		t.Fatal(err)
	}
	var expecting int64 = 0
	if i64 != expecting {
		t.Fatalf("Expected: %d Got: %d", expecting, i64)
	}
}


longhand.go
package longhand

import (
	"errors"
	"strings"
)

const maxInt32 = 1<<31 - 1
const minInt32 = -1 << 31

var numerals = map[string]int64{
	"zero":      0,
	"one":       1,
	"two":       2,
	"three":     3,
	"four":      4,
	"five":      5,
	"six":       6,
	"seven":     7,
	"eight":     8,
	"nine":      9,
	"ten":       10,
	"eleven":    11,
	"twelve":    12,
	"thirteen":  13,
	"fourteen":  14,
	"fifteen":   15,
	"sixteen":   16,
	"seventeen": 17,
	"eighteen":  18,
	"nineteen":  19,
	"twenty":    20,
	"thirty":    30,
	"forty":     40,
	"fifty":     50,
	"sixty":     60,
	"seventy":   70,
	"eighty":    80,
	"ninety":    90,
}

// Exclusive of 10
var ordersOfMagnitude = map[string]int64{
	"hundred":     100,
	"thousand":    1000,
	"k":           1000,
	"million":     1000000,
	"mil":         1000000,
	"m":           1000000,
	"billion":     1000000000,
	"bil":         1000000000,
	"b":           1000000000,
	"trillion":    1000000000000,
	"tril":        1000000000000,
	"t":           1000000000000,
	"quadrillion": 1000000000000000,
	"quad":        1000000000000000,
	"quintillion": 1000000000000000000,
	"quin":        1000000000000000000,
}

// LonghandError stores relevant information associated with a longhand error.
type LonghandError struct {
	Field   string // The infringing field
	Err     error  // The corresponding error
	FunName string // The associated function name
}

func (err *LonghandError) Error() string {
	return err.Err.Error() + ": \"" + err.Field + "\" in " + "longhand." + err.FunName
}

func rangeError(funName, s string) *LonghandError {
	return &LonghandError{s, errors.New("value out of range"), funName}
}

func syntaxError(funName, s string) *LonghandError {
	return &LonghandError{s, errors.New("invalid syntax"), funName}
}

func mul64(x int64, y int64) (int64, error) {
	const fnMul64 = "Mul64"
	z := x * y
	if y != 0 && x != z/y {
		return 0, errors.New("integer overflow")
	}
	return z, nil
}

// ParseLonghand64 interprets a longhand string lh and
// returns the corresponding 64-bit signed integer i64.
//
// If lh is empty or not in valid longhand, an error is
// returned and i64 = 0.
//
// The range for ParseLonghand64 is between -9,223,372,036,854,775,808
// and 9,223,372,036,854,775,807 (inclusive). If lh is not in this range
// an error is returned and i64 = 0.
//
// The errors returned by ParseLonghand64 are of type *LonghandError.
func ParseLonghand64(lh string) (i64 int64, err error) {
	const funName = "ParseLonghand64"

	if lh == "" {
		return 0, syntaxError(funName, lh)
	}

	var neg int64 = 1
	var preN, sum int64
	var in bool
	var b strings.Builder

	size := len(lh)
	for i := 0; i <= size; i++ {
		var c uint8
		if i < size {
			c = lh[i]
		}
		if c != ' ' && c != '\t' && c != '\r' && c != '\n' && c != '-' && i < size {
			in = true
			if 'A' <= c && c <= 'Z' {
				c += 32
			}
			b.WriteString(string(c))
		} else if in {
			// Handle current string in buffer
			s := b.String()
			// Validate s
			n, isN := numerals[s]
			m, isM := ordersOfMagnitude[s]
			if !isN && !isM {
				if s == "negative" {
					neg = -1
				} else if s != "and" {
					return 0, syntaxError(funName, s)
				}
			}
			// Process s
			if isN {
				if 0 < preN/10 && 0 < n/10 {
					sum *= 100
				} else if preN == n {
					sum *= 10
				}
				sum += n
				preN = n
			}
			if isM {
				if sum == 0 {
					sum += m
				} else {
					sum, err = mul64(sum, m)
					if err != nil {
						return 0, rangeError(funName, lh)
					}
				}
				if m != 100 {
					i64 += sum
					sum = 0
				}
				preN = 0
			}
			b.Reset()
			in = false
		}
	}

	i64 += sum
	i64 *= neg
	if neg == -1 && i64 > 0 || neg == 1 && i64 < 0 {
		return 0, rangeError(funName, lh)
	}
	return i64, nil
}

// ParseLonghand interprets a longhand string lh and
// returns the corresponding 32-bit signed integer i.
//
// If lh is empty or not in valid longhand, an error is
// returned and i = 0.
//
// The range for ParseLonghand is between -2,147,483,648
// and 2,147,483,647 (inclusive). If lh is not in this range
// an error is returned and i = 0.
//
// The errors returned by ParseLonghand are of type *LonghandError.
//
// ParseLonghand is equivalent to calling ParseLonghand64,
// enforcing 32-bit signed integer range, and performing an
// int type conversion.
func ParseLonghand(lh string) (i int, err error) {
	const funName = "ParseLonghand"

	i64, err := ParseLonghand64(lh)
	if e, ok := err.(*LonghandError); ok {
		e.FunName = funName
	}

	if minInt32 > i64 || i64 > maxInt32 {
		return 0, rangeError(funName, lh)
	}
	return int(i64), err
}
